crates/editor/src/editor.rs

```diff
@@ -9582,16 +9585,47 @@ impl EditorSnapshot {
     ) -> GutterDimensions {
         if self.show_gutter {
             let descent = cx.text_system().descent(font_id, font_size);
-            let gutter_padding_factor = 4.0;
-            let gutter_padding = (em_width * gutter_padding_factor).round();
-            // Avoid flicker-like gutter resizes when the line number gains another digit and only resize the gutter on files with N*10^5 lines.
-            let min_width_for_number_on_gutter = em_width * 4.0;
-            let gutter_width =
-                max_line_number_width.max(min_width_for_number_on_gutter) + gutter_padding * 2.0;
+
+            let show_git_gutter = matches!(
+                ProjectSettings::get_global(cx).git.git_gutter,
+                Some(GitGutterSetting::TrackedFiles)
+            );
+            let gutter_settings = EditorSettings::get_global(cx).gutter;
+
+            let line_gutter_width = if gutter_settings.line_numbers {
+                // Avoid flicker-like gutter resizes when the line number gains another digit and only resize the gutter on files with N*10^5 lines.
+                let min_width_for_number_on_gutter = em_width * 4.0;
+                max_line_number_width.max(min_width_for_number_on_gutter)
+            } else {
+                0.0.into()
+            };
+
+            let left_gutter_padding = match (
+                show_git_gutter,
+                gutter_settings.code_actions,
+                gutter_settings.line_numbers,
+            ) {
+                (true, false, false) => em_width,
+                (false, false, true) => em_width * 2.0,
+                (true, true, _) => em_width * 4.0,
+                (true, false, _) => em_width * 2.0,
+                (false, true, _) => em_width * 4.0,
+                _ => 0.0.into(),
+            };
+
+            let right_gutter_padding = match (gutter_settings.line_numbers, gutter_settings.folds) {
+                (true, true) => em_width * 4.0,
+                (true, false) => em_width * 2.0,
+                (false, true) => em_width * 3.0,
+                (false, false) => 0.0.into(),
+            };
```

Comment:
Let me know if these match expressions look hard to reason about, I couldn't think of a better way of getting the right paddings for each combination.

Reply:
Thanks for calling this out. I had nearly left a comment to this effect on the previous PR.

I think the match statements are OK, but it does worry me that we have to calculate what's going on in two places: once here, and once when actually rendering. Because most people will only test with one set of settings, it will be easier for things to regress.

I'd love to pair with you on doing a little refactoring to improve this: https://calendly.com/conradirwin/pairing, but I'll leave notes here in case you want to tackle on your own:

* Lets preserve the `GutterDimensions` struct inside `Layout` instead of pulling out the fields we care about.
* (Similarly lets pass it to `layout_blocks` and `BlockContext` instead of pulling out each field)
* Let's add methods to the `GutterDimensions` struct that take a `cx` and return an `Option<Pixels>` for these things:
*  `gutter_dimensions.line_number_right_padding(cx)` that return `None` if line numbers are disabled or `Some(right_padding)` if they are enabled.
* `gutter_dimensions.fold_right_padding(cx)`, ditto.

The goal here is to ensure that if we have line numbers disabled, it's not possible to calculate where to draw them, so we can be sure that positioning logic bugs can be tracked down to code inside `GutterDimensions`. 
(It does mean that in the case of logic bugs we may spend extra cycles computing things that we're never going to render; but that seems a better failure mode).

I am not sure, but I hope that if we did this consistently, we'd be able to replace these match statements with something that uses those methods to clarify the intent:

```
let padding_right = self.line_number_distance_from_right(cx).unwrap_or(self.folds_right_padding(cx)).unwrap_or(px(0.0))

let mut padding_left = self.code_actions_width(cx).unwrap_or(self.git_gutter_width(cx)).unwrap_or(px(0.0))
if settings.line_numbers && padding_left < em_width * 2.0 {
  padding_left = em_width * 2.0
}
```

Open to other ideas, but I do think we need to do at least one pass of cleanup on this.

Reply:
Alright, I got rid of the match expressions and replaced them with something similar to what you suggested. The visuals are not quite the same (there is a small difference in the right padding when line numbers and folds are on), but it still looks good to me and the code is much more understandable.

I'm not sure if the methods in GutterDimensions are necessary, since they would only be called in one place, but maybe I misunderstood.

If there is anything else we could change let me know and we could do the pair session next week (I haven't had the time because there's some construction going on in my house) :)

Reply:
@felipeagc sorry for leading you down the garden path there! The optional unwrapping ended up being much noisier than intended.

I think the fold indicators looked a bit tight in the version you had; and I realize that the code actions indicator was too big, so I've pushed some clean up to this branch, and will merge.